### PR TITLE
Product Collection: fix "Reset All" button in Editor filters

### LINKED ISSUE #46156: Product Collection - "Reset All" resets also the collection filters
**Background:**

Collections have predefined filters that should not be changed by merchants, hence they're hidden from Inspector Controls. However, when we use "Reset All" button, they're reset as well.

**Steps to reproduce:**

1. Create new post
2. Insert Product Collection
3. Choose On Sale collection - only On Sale products are displayed
4. Open Inspector Controls and change some filters, for example, Price Range. Set the value that limits the number of products, value depends on your set of products.
5. Open Filters three-dot menu and click "Reset All"

image

Expected: Products return to the state from step 3 (only On Sale products are displayed)
Actual: All the filters are reset and different set of products are displayed.

### PR DESCRIPTION
**Pull request by kmanijak**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

The "Reset All" button in Product Collection Editor filters worked incorrectly: it reset the filters which should not change for specific collections. For example, On Sale collection has an `onSale` filter enabled all the time (it's unchangeable by the merchant). However, using "Reset All" resets an unchangeable filter, in this case, `onSale`.

After checking it occurred to me we were doing a reset incorrectly by principle. This PR:
- removes the original reset by directly editing the query.
- adds `deselectCallback` to each filter which is then passed to both:
    - `onDeselect` - fired on deselecting specific filter,
    - `resetAllFilter` - fired when clicking "Reset All".
That leverages the native Gutenberg feature and also if the specific filter is out of control by the merchant, this won't be reset on clicking "Reset All".

Closes https://github.com/woocommerce/woocommerce/issues/46156

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create new post
2. Add Product Collection block
3. Choose On Sale collection - please note only on sale products are displayed
4. Apply two additional filters that would limit the number of displayed products, for example:
  - Price Range: max price (value depends on your local setup, choose a value that would filter some of the products out)
  - Taxonomies - Product category: (value depends on your local setup, choose a value that would filter some of the products out)
5. Click three-dot Filter menu and click "Reset All"

image

7. **Expected:** collection is back at the state from step 3: all the applied filters are reset and the collection still displays on sale products.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Review by albarin — APPROVED** [review]
👍

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/on-sale-control.tsx [hunk 1]
```diff
@@ -17,16 +17,17 @@ import { QueryControlProps } from '../../types';
 const OnSaleControl = ( props: QueryControlProps ) => {
 	const { query, setQueryAttribute } = props;
 
+	const deselectCallback = () => {
+		setQueryAttribute( { woocommerceOnSale: false } );
+	};
```
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/on-sale-control.tsx]: Nit: Maybe we can consider using default attributes constant here, i.e., `DEFAULT_ATTRIBUTES.query?.woocommerceOnSale`. This way, we have a single source of truth for all default values. What do you think?

Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/on-sale-control.tsx]: This applies for all the controls

**Review by imanish003 — APPROVED** [review]
Good work! Thank you for addressing the issue. I left a minor comment, but pre-approving the PR 🚀

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/on-sale-control.tsx (hunk 1 above)
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/on-sale-control.tsx]: Good idea! I used `DEFAULT_FILTERS` and cleaned that up a bit so it relies on `DEFAULT_QUERY`: https://github.com/woocommerce/woocommerce/pull/46226/commits/aaca6cfc20b0eaffb515a653431b84352ad6a847.

Thanks for this suggestion!

**Pull request merged**